File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/command/ProgramAlertsCommand.java
package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.base.CommandType;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.base.NonceEnabledCommand;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.base.builder.NonceEnabledCommandBuilder;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.definition.AlertConfiguration;

public final class ProgramAlertsCommand extends NonceEnabledCommand {
    private final List<AlertConfiguration> alertConfigurations;

    ProgramAlertsCommand(int uniqueId, short sequenceNumber, boolean multiCommandFlag, List<AlertConfiguration> alertConfigurations, int nonce) {
        super(CommandType.PROGRAM_ALERTS, uniqueId, sequenceNumber, multiCommandFlag, nonce);
        this.alertConfigurations = new ArrayList<>(alertConfigurations);
    }

    @Override public byte[] getEncoded() {
        ByteBuffer byteBuffer = ByteBuffer.allocate(getLength() + HEADER_LENGTH) //
                .put(encodeHeader(uniqueId, sequenceNumber, getLength(), multiCommandFlag)) //
                .put(commandType.getValue()) //
                .put(getBodyLength()) //
                .putInt(nonce);
        for (AlertConfiguration configuration : alertConfigurations) {
            byteBuffer.put(configuration.getEncoded());
        }
        return appendCrc(byteBuffer.array());
    }

    private short getLength() {
        return (short) (alertConfigurations.size() * 6 + 6);
    }

    private byte getBodyLength() {
        return (byte) (alertConfigurations.size() * 6 + 4);
    }

    @Override public String toString() {
        return "ProgramAlertsCommand{" +
                "alertConfigurations=" + alertConfigurations +
                ", nonce=" + nonce +
                ", commandType=" + commandType +
                ", uniqueId=" + uniqueId +
                ", sequenceNumber=" + sequenceNumber +
                ", multiCommandFlag=" + multiCommandFlag +
                '}';
    }

    public static final class Builder extends NonceEnabledCommandBuilder<Builder, ProgramAlertsCommand> {
        private List<AlertConfiguration> alertConfigurations;

        public Builder setAlertConfigurations(List<AlertConfiguration> alertConfigurations) {
            this.alertConfigurations = alertConfigurations;
            return this;
        }

        @Override protected final ProgramAlertsCommand buildCommand() {
            if (this.alertConfigurations == null) {
                throw new IllegalArgumentException("alertConfigurations can not be null");
            }
            return new ProgramAlertsCommand(uniqueId, sequenceNumber, multiCommandFlag, alertConfigurations, nonce);
        }
    }
}

File: pump/omnipod-dash/src/main/kotlin/app/aaps/pump/omnipod/dash/driver/pod/command/ProgramAlertsCommand.kt
package app.aaps.pump.omnipod.dash.driver.pod.command

import app.aaps.pump.omnipod.dash.driver.pod.command.base.CommandType
import app.aaps.pump.omnipod.dash.driver.pod.command.base.NonceEnabledCommand
import app.aaps.pump.omnipod.dash.driver.pod.command.base.builder.NonceEnabledCommandBuilder
import app.aaps.pump.omnipod.dash.driver.pod.definition.AlertConfiguration
import java.nio.ByteBuffer
import java.util.*

class ProgramAlertsCommand private constructor(
    uniqueId: Int,
    sequenceNumber: Short,
    multiCommandFlag: Boolean,
    alertConfigurations: List<AlertConfiguration>,
    nonce: Int
) : NonceEnabledCommand(CommandType.PROGRAM_ALERTS, uniqueId, sequenceNumber, multiCommandFlag, nonce) {

    private val alertConfigurations: List<AlertConfiguration>

    private fun getLength(): Short {
        return (alertConfigurations.size * 6 + 6).toShort()
    }

    private fun getBodyLength(): Byte {
        return (alertConfigurations.size * 6 + 4).toByte()
    }

    override val encoded: ByteArray
        get() {
            val byteBuffer: ByteBuffer = ByteBuffer.allocate(getLength() + HEADER_LENGTH)
                .put(encodeHeader(uniqueId, sequenceNumber, getLength(), multiCommandFlag))
                .put(commandType.value)
                .put(getBodyLength())
                .putInt(nonce)
            for (configuration in alertConfigurations) {
                byteBuffer.put(configuration.encoded)
            }
            return appendCrc(byteBuffer.array())
        }

    val encodedWithoutHeaderAndCRC32: ByteArray
        get() {
            val byteBuffer: ByteBuffer = ByteBuffer.allocate(getLength().toInt())
                .put(commandType.value)
                .put(getBodyLength())
                .putInt(nonce)
            for (configuration in alertConfigurations) {
                byteBuffer.put(configuration.encoded)
            }
            return byteBuffer.array()
        }

    override fun toString(): String {
        return "ProgramAlertsCommand{" +
            "alertConfigurations=" + alertConfigurations +
            ", nonce=" + nonce +
            ", commandType=" + commandType +
            ", uniqueId=" + uniqueId +
            ", sequenceNumber=" + sequenceNumber +
            ", multiCommandFlag=" + multiCommandFlag +
            '}'
    }

    class Builder : NonceEnabledCommandBuilder<Builder, ProgramAlertsCommand>() {

        private var alertConfigurations: List<AlertConfiguration>? = null

        fun setAlertConfigurations(alertConfigurations: List<AlertConfiguration>): Builder {
            this.alertConfigurations = alertConfigurations
            return this
        }

        override fun buildCommand(): ProgramAlertsCommand {
            requireNotNull(alertConfigurations) { "alertConfigurations can not be null" }
            return ProgramAlertsCommand(uniqueId!!, sequenceNumber!!, multiCommandFlag, alertConfigurations!!, nonce!!)
        }
    }

    init {
        this.alertConfigurations = ArrayList(alertConfigurations)
    }
}
